File: code/Project_4/BT6/python/tree_edit_distance_bb.py
import heapq

class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

def read_tree(filename):
    try:
        with open(filename, 'r') as f:
            n = int(f.readline().strip())
            if n == 0:
                return None
            edges = []
            for _ in range(n):
                line = list(map(int, f.readline().strip().split()))
                edges.append(line)
        return build_tree_from_edges(edges)
    except:
        return None

def build_tree_from_edges(edges):
    if not edges:
        return None
    
    nodes = {}
    children_set = set()
    
    for parent, left, right in edges:
        if parent not in nodes:
            nodes[parent] = TreeNode(parent)
        if left != -1:
            if left not in nodes:
                nodes[left] = TreeNode(left)
            nodes[parent].left = nodes[left]
            children_set.add(left)
        if right != -1:
            if right not in nodes:
                nodes[right] = TreeNode(right)
            nodes[parent].right = nodes[right]
            children_set.add(right)
    
    # Find root
    root = None
    for val in nodes:
        if val not in children_set:
            root = nodes[val]
            break
    
    return root

def tree_size(node):
    if not node:
        return 0
    return 1 + tree_size(node.left) + tree_size(node.right)

def lower_bound(T1, T2):
    """Simple lower bound estimation"""
    size1 = tree_size(T1)
    size2 = tree_size(T2)
    return abs(size1 - size2)

def ted_branch_bound(T1, T2):
    """Branch and Bound approach"""
    if not T1 and not T2:
        return 0
    if not T1:
        return tree_size(T2)
    if not T2:
        return tree_size(T1)
    
    # Priority queue: (cost, T1, T2)
    pq = [(0, T1, T2)]
    best_cost = float('inf')
    
    while pq:
        current_cost, t1, t2 = heapq.heappop(pq)
        
        if current_cost >= best_cost:
            continue
            
        if not t1 and not t2:
            best_cost = min(best_cost, current_cost)
            continue
        if not t1:
            best_cost = min(best_cost, current_cost + tree_size(t2))
            continue
        if not t2:
            best_cost = min(best_cost, current_cost + tree_size(t1))
            continue
        
        # Three operations with bounds
        lb = lower_bound(t1, t2)
        
        # Relabel
        relabel_cost = 0 if t1.val == t2.val else 1
        heapq.heappush(pq, (current_cost + relabel_cost, t1.left, t2.left))
        heapq.heappush(pq, (current_cost + relabel_cost, t1.right, t2.right))
        
        # Delete
        heapq.heappush(pq, (current_cost + 1, t1.left, t2))
        heapq.heappush(pq, (current_cost + 1, t1.right, t2))
        
        # Insert
        heapq.heappush(pq, (current_cost + 1, t1, t2.left))
        heapq.heappush(pq, (current_cost + 1, t1, t2.right))
    
    return best_cost

def main():
    T1 = read_tree('../tree1.inp')
    T2 = read_tree('../tree2.inp')
    
    distance = ted_branch_bound(T1, T2)
    print(f"Branch & Bound - Tree Edit Distance: {distance}")
    print("Done")

if __name__ == "__main__":
    main()

File: code/Project_4/BT6/c++/tree_edit_distance_bb.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <queue>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

struct State {
    int cost;
    TreeNode* t1;
    TreeNode* t2;
    
    bool operator>(const State& other) const {
        return cost > other.cost;
    }
};

TreeNode* read_tree(const string& filename) {
    ifstream infile(filename);
    if (!infile.is_open()) {
        return nullptr;
    }
    
    int n;
    infile >> n;
    if (n == 0) return nullptr;
    
    map<int, TreeNode*> nodes;
    set<int> children;
    
    for (int i = 0; i < n; i++) {
        int parent, left, right;
        infile >> parent >> left >> right;
        
        if (nodes.find(parent) == nodes.end())
            nodes[parent] = new TreeNode(parent);
            
        if (left != -1) {
            if (nodes.find(left) == nodes.end())
                nodes[left] = new TreeNode(left);
            nodes[parent]->left = nodes[left];
            children.insert(left);
        }
        
        if (right != -1) {
            if (nodes.find(right) == nodes.end())
                nodes[right] = new TreeNode(right);
            nodes[parent]->right = nodes[right];
            children.insert(right);
        }
    }
    
    // Find root
    for (auto& pair : nodes) {
        if (children.find(pair.first) == children.end()) {
            return pair.second;
        }
    }
    return nullptr;
}

int tree_size(TreeNode* node) {
    if (!node) return 0;
    return 1 + tree_size(node->left) + tree_size(node->right);
}

int lower_bound(TreeNode* T1, TreeNode* T2) {
    int size1 = tree_size(T1);
    int size2 = tree_size(T2);
    return abs(size1 - size2);
}

int ted_branch_bound(TreeNode* T1, TreeNode* T2) {
    if (!T1 && !T2) return 0;
    if (!T1) return tree_size(T2);
    if (!T2) return tree_size(T1);
    
    priority_queue<State, vector<State>, greater<State>> pq;
    int best_cost = INT_MAX;
    
    pq.push({0, T1, T2});
    
    while (!pq.empty()) {
        State current = pq.top();
        pq.pop();
        
        if (current.cost >= best_cost) continue;
        
        TreeNode* t1 = current.t1;
        TreeNode* t2 = current.t2;
        
        if (!t1 && !t2) {
            best_cost = min(best_cost, current.cost);
            continue;
        }
        if (!t1) {
            best_cost = min(best_cost, current.cost + tree_size(t2));
            continue;
        }
        if (!t2) {
            best_cost = min(best_cost, current.cost + tree_size(t1));
            continue;
        }
        
        // Three operations
        int relabel_cost = (t1->val == t2->val) ? 0 : 1;
        pq.push({current.cost + relabel_cost, t1->left, t2->left});
        pq.push({current.cost + relabel_cost, t1->right, t2->right});
        
        pq.push({current.cost + 1, t1->left, t2});
        pq.push({current.cost + 1, t1->right, t2});
        
        pq.push({current.cost + 1, t1, t2->left});
        pq.push({current.cost + 1, t1, t2->right});
    }
    
    return best_cost;
}

int main() {
    TreeNode* T1 = read_tree("../tree1.inp");
    TreeNode* T2 = read_tree("../tree2.inp");
    
    int distance = ted_branch_bound(T1, T2);
    cout << "Branch & Bound - Tree Edit Distance: " << distance << endl;
    cout << "Done" << endl;
    
    return 0;
}